File: src/lib/sabflow/actions/mailchimp.ts

'use server';

import crypto from 'crypto';

async function mailchimpFetch(apiKey: string, server: string, method: string, path: string, body?: any, logger?: any) {
    logger?.log(`[Mailchimp] ${method} ${path}`);
    const base64Auth = Buffer.from(`anystring:${apiKey}`).toString('base64');
    const url = `https://${server}.api.mailchimp.com/3.0${path}`;
    const options: RequestInit = {
        method,
        headers: {
            Authorization: `Basic ${base64Auth}`,
            'Content-Type': 'application/json',
        },
    };
    if (body !== undefined) options.body = JSON.stringify(body);
    const res = await fetch(url, options);
    if (res.status === 204) return {};
    const data = await res.json();
    if (!res.ok) {
        throw new Error(data?.detail || data?.title || `Mailchimp API error: ${res.status}`);
    }
    return data;
}

function md5Email(email: string): string {
    return crypto.createHash('md5').update(email.toLowerCase()).digest('hex');
}

function getServer(apiKey: string): string {
    return apiKey.split('-').pop() ?? 'us1';
}

export async function executeMailchimpAction(actionName: string, inputs: any, user: any, logger: any) {
    try {
        const apiKey = String(inputs.apiKey ?? '').trim();
        if (!apiKey) throw new Error('apiKey is required.');
        const server = getServer(apiKey);
        const mc = (method: string, path: string, body?: any) => mailchimpFetch(apiKey, server, method, path, body, logger);

        switch (actionName) {
            case 'addSubscriber': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                const firstName = String(inputs.firstName ?? '').trim();
                const lastName = String(inputs.lastName ?? '').trim();
                const status = String(inputs.status ?? 'subscribed').trim();
                if (!listId || !email) throw new Error('listId and email are required.');
                const body: any = { email_address: email, status };
                if (firstName || lastName) body.merge_fields = { FNAME: firstName, LNAME: lastName };
                const data = await mc('POST', `/lists/${listId}/members`, body);
                return { output: { id: data.id, email: data.email_address, status: data.status } };
            }

            case 'updateSubscriber': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                if (!listId || !email) throw new Error('listId and email are required.');
                const hash = md5Email(email);
                const body: any = {};
                if (inputs.status) body.status = String(inputs.status);
                if (inputs.firstName || inputs.lastName) {
                    body.merge_fields = { FNAME: String(inputs.firstName ?? ''), LNAME: String(inputs.lastName ?? '') };
                }
                const data = await mc('PATCH', `/lists/${listId}/members/${hash}`, body);
                return { output: { id: data.id, email: data.email_address, status: data.status } };
            }

            case 'getSubscriber': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                if (!listId || !email) throw new Error('listId and email are required.');
                const hash = md5Email(email);
                const data = await mc('GET', `/lists/${listId}/members/${hash}`);
                return { output: { id: data.id, email: data.email_address, status: data.status, tags: data.tags ?? [] } };
            }

            case 'archiveSubscriber': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                if (!listId || !email) throw new Error('listId and email are required.');
                const hash = md5Email(email);
                await mc('DELETE', `/lists/${listId}/members/${hash}`);
                return { output: { archived: 'true', email } };
            }

            case 'addTag': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                const tagName = String(inputs.tagName ?? '').trim();
                if (!listId || !email || !tagName) throw new Error('listId, email, and tagName are required.');
                const hash = md5Email(email);
                await mc('POST', `/lists/${listId}/members/${hash}/tags`, { tags: [{ name: tagName, status: 'active' }] });
                return { output: { tagged: 'true', tagName } };
            }

            case 'removeTag': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                const tagName = String(inputs.tagName ?? '').trim();
                if (!listId || !email || !tagName) throw new Error('listId, email, and tagName are required.');
                const hash = md5Email(email);
                await mc('POST', `/lists/${listId}/members/${hash}/tags`, { tags: [{ name: tagName, status: 'inactive' }] });
                return { output: { removed: 'true', tagName } };
            }

            case 'listMembers': {
                const listId = String(inputs.listId ?? '').trim();
                const count = Number(inputs.count ?? 100);
                const status = String(inputs.status ?? '').trim();
                if (!listId) throw new Error('listId is required.');
                let path = `/lists/${listId}/members?count=${count}`;
                if (status) path += `&status=${status}`;
                const data = await mc('GET', path);
                return { output: { members: data.members ?? [], total: String(data.total_items ?? 0) } };
            }

            case 'createCampaign': {
                const listId = String(inputs.listId ?? '').trim();
                const subject = String(inputs.subject ?? '').trim();
                const fromName = String(inputs.fromName ?? '').trim();
                const replyTo = String(inputs.replyTo ?? '').trim();
                const title = String(inputs.title ?? '').trim();
                if (!listId || !subject || !fromName || !replyTo) throw new Error('listId, subject, fromName, and replyTo are required.');
                const data = await mc('POST', '/campaigns', {
                    type: 'regular',
                    recipients: { list_id: listId },
                    settings: { subject_line: subject, from_name: fromName, reply_to: replyTo, title: title || subject },
                });
                return { output: { id: data.id, status: data.status, webId: String(data.web_id ?? '') } };
            }

            case 'sendCampaign': {
                const campaignId = String(inputs.campaignId ?? '').trim();
                if (!campaignId) throw new Error('campaignId is required.');
                await mc('POST', `/campaigns/${campaignId}/actions/send`);
                return { output: { sent: 'true', campaignId } };
            }

            case 'getCampaignReport': {
                const campaignId = String(inputs.campaignId ?? '').trim();
                if (!campaignId) throw new Error('campaignId is required.');
                const data = await mc('GET', `/reports/${campaignId}`);
                return { output: { opens: String(data.opens?.unique_opens ?? 0), clicks: String(data.clicks?.unique_clicks ?? 0), bounces: String(data.bounces?.hard_bounces ?? 0), unsubscribes: String(data.unsubscribes ?? 0) } };
            }

            case 'createTag': {
                const listId = String(inputs.listId ?? '').trim();
                const tagName = String(inputs.tagName ?? '').trim();
                if (!listId || !tagName) throw new Error('listId and tagName are required.');
                const data = await mc('POST', `/lists/${listId}/segments`, { name: tagName, static_segment: [], options: {} });
                return { output: { id: String(data.id), name: data.name } };
            }

            case 'addSubscriberNote': {
                const listId = String(inputs.listId ?? '').trim();
                const email = String(inputs.email ?? '').trim();
                const note = String(inputs.note ?? '').trim();
                if (!listId || !email || !note) throw new Error('listId, email, and note are required.');
                const hash = md5Email(email);
                const data = await mc('POST', `/lists/${listId}/members/${hash}/notes`, { note });
                return { output: { id: String(data.note_id), note: data.note } };
            }

            default:
                return { error: `Mailchimp action "${actionName}" is not implemented.` };
        }
    } catch (e: any) {
        return { error: e.message || 'Mailchimp action failed.' };
    }
}

File: rust/crates/sabflow-nodes/src/nodes/mailchimp.rs
//! Mailchimp node.
//!
//! Implements list, member, and campaign operations against the Mailchimp
//! Marketing API v3.0. The API key is suffixed with the data-center identifier
//! (e.g. `abc123-us21`); the portion after the dash is used as the host prefix
//! so the base URL becomes `https://{dc}.api.mailchimp.com/3.0`.
//!
//! Authenticates with HTTP Basic where the username is any non-empty string
//! and the password is the API key.

use async_trait::async_trait;
use md5::{Digest, Md5};
use serde_json::{Map, Value, json};

use crate::{
    context::{ExecutionContext, NodeInput, NodeOutput},
    descriptor::{
        CredentialBinding, NodeCategory, NodeDescriptor, NodeProperty, NodePropertyOption,
        NodePropertyType,
    },
    error::{NodeError, NodeResult},
    node::Node,
};

pub struct MailchimpNode;

fn opt(name: &str, value: &str) -> NodePropertyOption {
    NodePropertyOption {
        name: name.to_string(),
        value: json!(value),
        description: None,
    }
}

#[async_trait]
impl Node for MailchimpNode {
    fn descriptor(&self) -> NodeDescriptor {
        NodeDescriptor::new(
            "mailchimp",
            "Mailchimp",
            "Mailchimp list, member, and campaign operations",
            NodeCategory::Marketing,
        )
        .icon("send")
        .color("#FFE01B")
        .credentials(vec![CredentialBinding {
            name: "mailchimpApi".into(),
            display_name: "Mailchimp API Key".into(),
            required: true,
        }])
        .properties(vec![
            NodeProperty::new("credentialId", "Credential", NodePropertyType::Credential)
                .required(),
            NodeProperty::new("resource", "Resource", NodePropertyType::Options)
                .options(vec![
                    opt("List", "list"),
                    opt("Member", "member"),
                    opt("Campaign", "campaign"),
                ])
                .default(json!("member"))
                .required(),
            // list operations
            NodeProperty::new("operation", "Operation", NodePropertyType::Options)
                .options(vec![opt("Get All", "getAll"), opt("Get", "get")])
                .default(json!("getAll"))
                .show_when("resource", &["list"])
                .required(),
            // member operations
            NodeProperty::new("operation", "Operation", NodePropertyType::Options)
                .options(vec![
                    opt("Add", "add"),
                    opt("Get", "get"),
                    opt("Update", "update"),
                    opt("Delete", "delete"),
                ])
                .default(json!("add"))
                .show_when("resource", &["member"])
                .required(),
            // campaign operations
            NodeProperty::new("operation", "Operation", NodePropertyType::Options)
                .options(vec![
                    opt("Get All", "getAll"),
                    opt("Get", "get"),
                    opt("Create", "create"),
                    opt("Send", "send"),
                ])
                .default(json!("getAll"))
                .show_when("resource", &["campaign"])
                .required(),
            // shared list id (member operations + list:get)
            NodeProperty::new("listId", "List ID", NodePropertyType::String)
                .placeholder("a1b2c3d4e5")
                .description("Mailchimp audience (list) id"),
            // member email
            NodeProperty::new("email", "Email", NodePropertyType::String)
                .placeholder("user@example.com")
                .show_when("operation", &["add", "get", "update", "delete"]),
            // member status
            NodeProperty::new("status", "Status", NodePropertyType::Options)
                .options(vec![
                    opt("Subscribed", "subscribed"),
                    opt("Unsubscribed", "unsubscribed"),
                    opt("Pending", "pending"),
                ])
                .default(json!("subscribed"))
                .show_when("operation", &["add", "update"]),
            // merge fields
            NodeProperty::new("mergeFields", "Merge Fields", NodePropertyType::Json)
                .show_when("operation", &["add", "update"])
                .description("JSON object of merge field values, e.g. {\"FNAME\":\"Jane\",\"LNAME\":\"Doe\"}"),
            // campaign id
            NodeProperty::new("campaignId", "Campaign ID", NodePropertyType::String)
                .show_when("operation", &["get", "send"]),
            // campaign creation
            NodeProperty::new("campaignType", "Campaign Type", NodePropertyType::Options)
                .options(vec![opt("Regular", "regular"), opt("Plaintext", "plaintext")])
                .default(json!("regular"))
                .show_when("operation", &["create"]),
            NodeProperty::new("subject", "Subject Line", NodePropertyType::String)
                .show_when("operation", &["create"]),
            NodeProperty::new("fromName", "From Name", NodePropertyType::String)
                .show_when("operation", &["create"]),
            NodeProperty::new("replyTo", "Reply-To Email", NodePropertyType::String)
                .show_when("operation", &["create"]),
        ])
    }

    async fn execute(
        &self,
        ctx: &mut ExecutionContext,
        _input: NodeInput,
        params: &Value,
    ) -> NodeResult<NodeOutput> {
        let cred_id = ctx.param_str(params, "credentialId")?;
        let cred = ctx.credential(&cred_id)?;
        let api_key = cred
            .data
            .get("apiKey")
            .ok_or_else(|| NodeError::MissingParameter("apiKey".into()))?
            .clone();

        let dc = api_key
            .rsplit('-')
            .next()
            .filter(|s| !s.is_empty() && *s != api_key)
            .ok_or_else(|| NodeError::InvalidParameter {
                name: "apiKey".into(),
                reason: "API key must end with a data-center suffix like '-us21'".into(),
            })?
            .to_string();
        let base = format!("https://{dc}.api.mailchimp.com/3.0");

        let resource = ctx
            .param_str_opt(params, "resource")
            .unwrap_or_else(|| "member".to_string());
        let operation = ctx.param_str(params, "operation")?;

        let body: Value = match (resource.as_str(), operation.as_str()) {
            // ----- Lists -----
            ("list", "getAll") => {
                send_request(ctx, &api_key, Method::Get, &format!("{base}/lists"), None).await?
            }
            ("list", "get") => {
                let list_id = ctx.param_str(params, "listId")?;
                let url = format!("{base}/lists/{list_id}");
                send_request(ctx, &api_key, Method::Get, &url, None).await?
            }
            // ----- Members -----
            ("member", "add") => {
                let list_id = ctx.param_str(params, "listId")?;
                let email = ctx.param_str(params, "email")?;
                let status = ctx
                    .param_str_opt(params, "status")
                    .unwrap_or_else(|| "subscribed".to_string());
                let mut payload = Map::new();
                payload.insert("email_address".into(), json!(email));
                payload.insert("status".into(), json!(status));
                if let Some(mf) = parse_json_param(ctx, params, "mergeFields") {
                    payload.insert("merge_fields".into(), mf);
                }
                let url = format!("{base}/lists/{list_id}/members");
                send_request(
                    ctx,
                    &api_key,
                    Method::Post,
                    &url,
                    Some(Value::Object(payload)),
                )
                .await?
            }
            ("member", "get") => {
                let list_id = ctx.param_str(params, "listId")?;
                let email = ctx.param_str(params, "email")?;
                let hash = subscriber_hash(&email);
                let url = format!("{base}/lists/{list_id}/members/{hash}");
                send_request(ctx, &api_key, Method::Get, &url, None).await?
            }
            ("member", "update") => {
                let list_id = ctx.param_str(params, "listId")?;
                let email = ctx.param_str(params, "email")?;
                let hash = subscriber_hash(&email);
                let status = ctx
                    .param_str_opt(params, "status")
                    .unwrap_or_else(|| "subscribed".to_string());
                let mut payload = Map::new();
                payload.insert("email_address".into(), json!(email));
                payload.insert("status".into(), json!(status));
                if let Some(mf) = parse_json_param(ctx, params, "mergeFields") {
                    payload.insert("merge_fields".into(), mf);
                }
                let url = format!("{base}/lists/{list_id}/members/{hash}");
                send_request(
                    ctx,
                    &api_key,
                    Method::Put,
                    &url,
                    Some(Value::Object(payload)),
                )
                .await?
            }
            ("member", "delete") => {
                let list_id = ctx.param_str(params, "listId")?;
                let email = ctx.param_str(params, "email")?;
                let hash = subscriber_hash(&email);
                let url = format!("{base}/lists/{list_id}/members/{hash}");
                send_request(ctx, &api_key, Method::Delete, &url, None).await?
            }
            // ----- Campaigns -----
            ("campaign", "getAll") => {
                let url = format!("{base}/campaigns");
                send_request(ctx, &api_key, Method::Get, &url, None).await?
            }
            ("campaign", "get") => {
                let campaign_id = ctx.param_str(params, "campaignId")?;
                let url = format!("{base}/campaigns/{campaign_id}");
                send_request(ctx, &api_key, Method::Get, &url, None).await?
            }
            ("campaign", "create") => {
                let list_id = ctx.param_str(params, "listId")?;
                let campaign_type = ctx
                    .param_str_opt(params, "campaignType")
                    .unwrap_or_else(|| "regular".to_string());
                let subject = ctx.param_str(params, "subject")?;
                let from_name = ctx.param_str(params, "fromName")?;
                let reply_to = ctx.param_str(params, "replyTo")?;
                let payload = json!({
                    "type": campaign_type,
                    "recipients": { "list_id": list_id },
                    "settings": {
                        "subject_line": subject,
                        "from_name": from_name,
                        "reply_to": reply_to,
                    },
                });
                let url = format!("{base}/campaigns");
                send_request(ctx, &api_key, Method::Post, &url, Some(payload)).await?
            }
            ("campaign", "send") => {
                let campaign_id = ctx.param_str(params, "campaignId")?;
                let url = format!("{base}/campaigns/{campaign_id}/actions/send");
                send_request(ctx, &api_key, Method::Post, &url, None).await?
            }
            (res, op) => {
                return Err(NodeError::InvalidParameter {
                    name: "operation".into(),
                    reason: format!("unknown {res} operation: {op}"),
                });
            }
        };

        Ok(NodeOutput::single(vec![body]))
    }
}

#[derive(Clone, Copy)]
enum Method {
    Get,
    Post,
    Put,
    Delete,
}

async fn send_request(
    ctx: &ExecutionContext,
    api_key: &str,
    method: Method,
    url: &str,
    payload: Option<Value>,
) -> NodeResult<Value> {
    let mut req = match method {
        Method::Get => ctx.http.get(url),
        Method::Post => ctx.http.post(url),
        Method::Put => ctx.http.put(url),
        Method::Delete => ctx.http.delete(url),
    };
    req = req
        .basic_auth("anystring", Some(api_key))
        .header("Content-Type", "application/json");
    if let Some(body) = payload {
        req = req.json(&body);
    }
    let res = req.send().await?;
    let status = res.status();
    let text = res.text().await.unwrap_or_default();
    let json_body: Value = if text.is_empty() {
        Value::Null
    } else {
        serde_json::from_str(&text).unwrap_or(Value::String(text.clone()))
    };
    if !status.is_success() {
        return Err(NodeError::UpstreamError {
            status: status.as_u16(),
            body: json_body.to_string(),
        });
    }
    Ok(json_body)
}

/// Mailchimp's "subscriber hash" — the MD5 (lowercase hex) of the lowercased
/// email address.
fn subscriber_hash(email: &str) -> String {
    let mut hasher = Md5::new();
    hasher.update(email.to_lowercase().as_bytes());
    let digest = hasher.finalize();
    let mut out = String::with_capacity(32);
    for byte in digest {
        out.push_str(&format!("{byte:02x}"));
    }
    out
}

fn parse_json_param(ctx: &ExecutionContext, params: &Value, key: &str) -> Option<Value> {
    let raw = params.get(key)?;
    match raw {
        Value::Null => None,
        Value::String(s) => {
            let s = ctx.substitute(s);
            let trimmed = s.trim();
            if trimmed.is_empty() {
                None
            } else {
                serde_json::from_str::<Value>(trimmed).ok()
            }
        }
        other => Some(substitute_value(ctx, other.clone())),
    }
}

fn substitute_value(ctx: &ExecutionContext, v: Value) -> Value {
    match v {
        Value::String(s) => Value::String(ctx.substitute(&s)),
        Value::Array(arr) => {
            Value::Array(arr.into_iter().map(|x| substitute_value(ctx, x)).collect())
        }
        Value::Object(map) => {
            let mut out = Map::new();
            for (k, val) in map {
                out.insert(k, substitute_value(ctx, val));
            }
            Value::Object(out)
        }
        other => other,
    }
}
